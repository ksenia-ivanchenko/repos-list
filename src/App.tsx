import { FC, useCallback, useEffect, useState } from "react";
import { useDispatch, useSelector } from "./store";
import { clearRepos, getRepos } from "./store/slices";
import { RepoItem } from "./components";
import { LoadingOutlined } from "@ant-design/icons";
import { Select, Spin } from "antd";
import { TSortParams } from "./store/slices/repo/types";
import styles from "./app.module.css";
import "./styles/reset.module.css";

export const App: FC = () => {
    const dispatch = useDispatch();
    const { repos, loading, totalCount, requestError } = useSelector(
        (state) => state.repo
    );
    const [currentPage, setCurrentPage] = useState(1);
    const [getNewRepos, setGetNewRepos] = useState(true);
    const [sortValue, setSortValue] = useState<TSortParams>("forks");

    useEffect(() => {
        if (getNewRepos) {
            dispatch(getRepos({ page: currentPage, sortQuery: sortValue }))
                .then(() => {
                    setCurrentPage((prev) => prev + 1);
                })
                .finally(() => {
                    setGetNewRepos(false);
                });
        }
    }, [getNewRepos]);

    const handleSelectSort = (value: TSortParams) => {
        dispatch(clearRepos());
        setSortValue(value);
        setGetNewRepos(true);
        setCurrentPage(1);
    };

    const handleScroll = useCallback(
        (e) => {
            const needNewRepos =
                e.target.documentElement.scrollHeight -
                    (e.target.documentElement.scrollTop + window.innerHeight) <
                100;

            if (needNewRepos && repos.length < totalCount) {
                setGetNewRepos(true);
            }
        },
        [repos, totalCount]
    );

    useEffect(() => {
        document.addEventListener("scroll", handleScroll);

        return () => document.removeEventListener("scroll", handleScroll);
    }, [handleScroll]);

    return (
        <main className={styles.main}>
            <span>Отсортировать по: </span>
            <Select
                defaultValue="количеству форков"
                onChange={handleSelectSort}
                className={styles.select}
                options={[
                    { value: "forks", label: "количеству форков" },
                    { value: "stars", label: "количеству звезд" },
                    {
                        value: "updated",
                        label: "времени обновления",
                    },
                ]}
            />
            <ul className={styles.list}>
                {repos.map((repo) => (
                    <RepoItem repo={repo} key={repo.id}></RepoItem>
                ))}
                {loading && (
                    <Spin indicator={<LoadingOutlined spin />} size="large" />
                )}
                {requestError && <span>Что-то пошло не так...</span>}
            </ul>
        </main>
    );
};

export default App;
