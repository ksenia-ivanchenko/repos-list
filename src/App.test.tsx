import { screen, waitFor } from "@testing-library/react";
import App from "./App";
import { renderWithRedux } from "./tests/helpers/renderWithRedux";
import { mockGetRepo } from "./tests/mock/mockGetRepos";

describe("App", () => {
    it("should create App", () => {
        const component = renderWithRedux(<App />);
        expect(component).toMatchSnapshot();
    });

    it("should render App correctly with default state", () => {
        renderWithRedux(<App />, {
            repo: {
                repos: [],
                totalCount: 0,
                loading: false,
                requestError: null,
            },
        });

        expect(screen.getByText(/отсортировать по:/i)).toBeInTheDocument();
    });

    it("should render repos after fetching", async () => {
        renderWithRedux(<App />, {
            repo: {
                repos: mockGetRepo.items,
                totalCount: mockGetRepo.total_count,
                loading: false,
                requestError: null,
            },
        });

        await waitFor(() => {
            expect(screen.getByText(/bootstrap/i)).toBeInTheDocument();
        });
    });

    it("should show loading spinner when loading", async () => {
        renderWithRedux(<App />, {
            repo: {
                repos: [],
                totalCount: 0,
                loading: true,
                requestError: null,
            },
        });
        await waitFor(() =>
            expect(
                screen.getByRole("img", { name: /loading/i })
            ).toBeInTheDocument()
        );
    });

    it("should display error message if there is a request error", () => {
        renderWithRedux(<App />, {
            repo: {
                repos: [],
                totalCount: 0,
                loading: false,
                requestError: "Ошибка загрузки",
            },
        });

        expect(screen.getByText(/что-то пошло не так.../i)).toBeInTheDocument();
    });
});
