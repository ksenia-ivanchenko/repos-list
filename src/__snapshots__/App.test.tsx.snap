// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should create App 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="main"
      >
        <span>
          Отсортировать по: 
        </span>
        <div
          class="ant-select ant-select-outlined select css-dev-only-do-not-override-nqoqt9 ant-select-single ant-select-show-arrow"
        >
          <div
            class="ant-select-selector"
          >
            <span
              class="ant-select-selection-wrap"
            >
              <span
                class="ant-select-selection-search"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="ant-select-selection-search-input"
                  id="rc_select_TEST_OR_SSR"
                  readonly=""
                  role="combobox"
                  style="opacity: 0;"
                  type="search"
                  unselectable="on"
                  value=""
                />
              </span>
              <span
                class="ant-select-selection-item"
                title="количеству форков"
              >
                количеству форков
              </span>
            </span>
          </div>
          <span
            aria-hidden="true"
            class="ant-select-arrow"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              aria-label="down"
              class="anticon anticon-down ant-select-suffix"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="down"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                />
              </svg>
            </span>
          </span>
        </div>
        <ul
          class="list"
        >
          <div
            aria-busy="true"
            aria-live="polite"
            class="ant-spin ant-spin-lg ant-spin-spinning css-dev-only-do-not-override-nqoqt9"
          >
            <span
              aria-label="loading"
              class="anticon anticon-loading anticon-spin ant-spin-dot"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="loading"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="0 0 1024 1024"
                width="1em"
              >
                <path
                  d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
                />
              </svg>
            </span>
          </div>
        </ul>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="main"
    >
      <span>
        Отсортировать по: 
      </span>
      <div
        class="ant-select ant-select-outlined select css-dev-only-do-not-override-nqoqt9 ant-select-single ant-select-show-arrow"
      >
        <div
          class="ant-select-selector"
        >
          <span
            class="ant-select-selection-wrap"
          >
            <span
              class="ant-select-selection-search"
            >
              <input
                aria-autocomplete="list"
                aria-controls="rc_select_TEST_OR_SSR_list"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="rc_select_TEST_OR_SSR_list"
                autocomplete="off"
                class="ant-select-selection-search-input"
                id="rc_select_TEST_OR_SSR"
                readonly=""
                role="combobox"
                style="opacity: 0;"
                type="search"
                unselectable="on"
                value=""
              />
            </span>
            <span
              class="ant-select-selection-item"
              title="количеству форков"
            >
              количеству форков
            </span>
          </span>
        </div>
        <span
          aria-hidden="true"
          class="ant-select-arrow"
          style="user-select: none;"
          unselectable="on"
        >
          <span
            aria-label="down"
            class="anticon anticon-down ant-select-suffix"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
      <ul
        class="list"
      >
        <div
          aria-busy="true"
          aria-live="polite"
          class="ant-spin ant-spin-lg ant-spin-spinning css-dev-only-do-not-override-nqoqt9"
        >
          <span
            aria-label="loading"
            class="anticon anticon-loading anticon-spin ant-spin-dot"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="loading"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
              />
            </svg>
          </span>
        </div>
      </ul>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
